motionplan/nearestNeighbor.go

```diff
@@ -111,24 +107,21 @@ func (nm *neighborManager) parallelNearestNeighbor(
 	for k := range rrtMap {
 		nm.nnKeys <- k
 	}
-	nm.nnLock.Lock()
-	nm.ready = true
-	nm.nnLock.Unlock()
+	close(nm.nnKeys)
+
 	var best node
 	bestDist := math.Inf(1)
-	returned := 0
-	for returned < nm.nCPU {
-		select {
-		case nn := <-nm.neighbors:
-			returned++
-			if nn != nil {
-				// nn will be nil if the ctx is cancelled
-				if nn.dist < bestDist {
-					bestDist = nn.dist
-					best = nn.node
-				}
-			}
-		default:
+	for workerIdx := 0; workerIdx < nm.nCPU; workerIdx++ {
+		candidate := <-nm.neighbors
+		if candidate == nil {
+			// Seeing a `nil` here implies the workers did not get to all of the candidate
+			// neighbors. And thus we don't have the right answer to return.
+			return nil
```

Comment:
I think this can cause a panic as written.

It will return on the first `nil` it sees, triggering the `defer close(nm.neighbors)` above, and any remaining workers would panic on a write to a closed channel.

I think we need to drain `nm.neighbors` before returning.

Reply:
Ah, yes -- you are right. Thanks!